### PR TITLE
feat: Support issue IDs as canonical parameters

### PR DESCRIPTION
**Pull request by dcramer**

### CODE REVIEW

**Comment by dcramer** [comment]
This would need tests + migrations, but putting this up to see what others think.

It'd allow us to use FOO-1ZA as a canonical resource in API endpoints and URLs.

For example, https://sentry.io/sentry/issues/FOO-1ZA

**Comment by wedamija** [comment]
This seems ok to me. You'll need to change a bunch of url regexes for this to work though, we just search for `\d+` at the moment.

**Review by markstory — COMMENTED** [review]
I like the idea. It improves the ergonomics of our URLs. Would we also want to update the permalink in emails and API responses?

File: src/sentry/models/group.py
Comment (dcramer) [reply src/sentry/models/group.py]: Maybe if we used it anywhere else, but given it still has to query multiple ways it seems unneeded at this point

**Comment by dcramer** [comment]
Going to finish this off in some free cycles. I'm going to expose it in frontend routes too if its not too hard (but not use it as canonical yet).

**Comment by dcramer** [comment]
Its a chunk of work to make the URLs accept it in the UI right now as they rely on GroupStore which would need to be id + short id aware first. Gonna not do that yet, and we can worry about that in a follow up.

**Comment by dcramer** [comment]
Rebasing and going to merge this when I get time this week.

**Comment by dcramer** [comment]
@markstory @wedamija actually yall didnt sign off on this. Mind reviewing formally?

File: src/sentry/south_migrations/0489_auto__add_field_groupredirect_previous_short_id__add_field_groupredire.py
Comment (wedamija) [inline src/sentry/south_migrations/0489_auto__add_field_groupredirect_previous_short_id__add_field_groupredire.py]: This needs to be a unique index

**Review by wedamija — CHANGES_REQUESTED** [review]

File: src/sentry/api/urls.py
Comment (dcramer) [reply src/sentry/api/urls.py]: I can revert - noticed it along the way and brings it in line with most other endpoints.

Right now it'd be a bit awkward if you queried /events/foo/ and got a generic 404 vs an "event not found".

File: src/sentry/models/groupredirect.py
Comment (dcramer) [reply src/sentry/models/groupredirect.py]: I dont think we can, but I dont quite recall the internals

File: src/sentry/api/urls.py
Comment (wedamija) [reply src/sentry/api/urls.py]: Not blocking on this at all, mostly just curious about the change.

**Comment by dcramer** [comment]
Resolved the index concern - anyone see anything else blocking?

**Review by wedamija — APPROVED** [review]
Looks good. Migrations are up to 0490, so you just need to rebuild the state before merging.

**Pull request merged**
